File: 2023/go/day-04/day-04.go
package main

import (
	"fmt"
	utils "github.com/markusthielker/advent-of-code/2023/go"
	"strconv"
	"strings"
)

func main() {

	part1 := func(input []string) (result int) {

		for _, card := range input {

			parts := strings.Split(card, "|")
			winningNumbers := toIntSlice(strings.Fields(strings.Split(parts[0], ":")[1]))
			cardNumbers := toIntSlice(strings.Fields(parts[1]))

			wins := 0
			for _, num := range cardNumbers {
				if contains(winningNumbers, num) {
					wins++
				}
			}

			points := 0
			if wins > 0 {
				points = 1
				for i := 1; i < wins; i++ {
					points *= 2
				}
			}

			result += points
		}

		return
	}

	part2 := func(input []string) (result int) {

		// assign each card a count of 1
		cardCounts := make(map[int]int)
		for index := range input {
			cardCounts[index] = 1
		}

		// iterate over each card
		for cardNumber, card := range input {

			parts := strings.Split(card, "|")
			winningNumbers := toIntSlice(strings.Fields(strings.Split(parts[0], ":")[1]))
			cardNumbers := toIntSlice(strings.Fields(parts[1]))

			wins := 0
			for _, num := range cardNumbers {
				if contains(winningNumbers, num) {
					wins++
				}
			}

			for i := 0; i < wins; i++ {
				nextIndex := cardNumber + (i + 1)
				currentCardWins := cardCounts[cardNumber]
				cardCounts[nextIndex] = cardCounts[nextIndex] + currentCardWins
			}
		}

		return sumIntValues(cardCounts)
	}

	input := utils.ReadInput("day-04/input")

	fmt.Println(part1(input))
	fmt.Println(part2(input))
}

func toIntSlice(strings []string) []int {
	var ints []int
	for _, s := range strings {
		if s != "" {
			if num, err := strconv.Atoi(s); err == nil {
				ints = append(ints, num)
			}
		}
	}
	return ints
}

func contains(slice []int, val int) bool {
	for _, item := range slice {
		if item == val {
			return true
		}
	}
	return false
}

func sumIntValues(m map[int]int) (sum int) {
	sum = 0
	for _, v := range m {
		sum += v
	}
	return
}

File: 2023/kotlin/src/day-04/Day04.kt
fun main() {

    fun part1(input: List<String>): Int {

        var result = 0

        for (card in input) {

            val winningNumbers = parseWinningNumbers(card)
            val cardNumbers = parseCardNumbers(card)

            val wins = cardNumbers.filter { it in winningNumbers }.size

            var points = if (wins == 0) 0 else 1
            repeat(wins - 1) {
                points *= 2
            }

            result += points
        }

        return result
    }

    fun part2(input: List<String>): Int {

        // assign each card a count of 1
        val cardCounts = mutableMapOf<Int, Int>()
        input.forEachIndexed { index, _ -> cardCounts[index] = 1 }

        // iterate over each card
        for ((index, s) in input.withIndex()) {

            val winningNumbers = parseWinningNumbers(s)
            val cardNumbers = parseCardNumbers(s)

            val wins = cardNumbers.filter { it in winningNumbers }.size

            repeat(wins) {

                val nextIndex = index + (it + 1)
                val currentCardWins = cardCounts[index]!!

                cardCounts[nextIndex] = cardCounts[nextIndex]!! + currentCardWins
            }
        }

        return cardCounts.map { it.value }.sum()
    }

    val input = readInput("day-04/input")
    part1(input).println()
    part2(input).println()
}

fun parseWinningNumbers(card: String): List<Int> {
    return card
        .split(":")[1]
        .split("|")[0]
        .split(" ")
        .filter { it.isNotBlank() }
        .map { it.toInt() }
}

fun parseCardNumbers(card: String): List<Int> {
    return card
        .split("|")[1]
        .split(" ")
        .filter { it.isNotBlank() }
        .map { it.toInt() }
}
